src/main/java/uk/gov/nationalarchives/pdi/step/xml/CanonicalStep.java

```diff
@@ -0,0 +1,121 @@
+        super(stepMeta, stepDataInterface, copyNr, transMeta, trans);
+    }
+
+    @Override
+    public boolean processRow(final StepMetaInterface smi, final StepDataInterface sdi) throws KettleException {
+        final CanonicalStepMeta meta = (CanonicalStepMeta) smi;
+        final CanonicalStepData data = (CanonicalStepData) sdi;
+
+        final Object[] row = getRow();
+        if (row == null) {
+            setOutputDone();
+            return false;
+        }
+
+        if(first) {
+            first = false;
+            data.setOutputRowMeta(getInputRowMeta().clone());
+            meta.getFields( data.getOutputRowMeta(), getStepname(), null, null, this, null, null );
+            data.setXmlFieldIdx(getInputRowMeta().indexOfValue(meta.getInputField()));
+            data.setOutputFieldIndex(data.getOutputRowMeta().indexOfValue(meta.getOutputField()));
+            org.apache.xml.security.Init.init();
+            data.setCanonicalizer(getCanonicalizer());
+            data.setDocumentBuilder(getDocumentBuilder());
+        }
+        final Object xmlFieldValue = row[data.getXmlFieldIdx()];
+        if(xmlFieldValue instanceof String) {
+            final CanonicalizationResult result = process((String)xmlFieldValue,data);
+            if(result.hasErrors()) {
+                putError(data.getOutputRowMeta(), row, result.getErrorCount(), result.getErrorMessage(), meta.getInputField(), "CanonicalStep001");
+            } else {
+                Object[] outputRow = RowDataUtil.resizeArray(row,data.getOutputRowMeta().size());
+                outputRow[data.getOutputFieldIndex()] = result.getCanonicalXml();
+                putRow(data.getOutputRowMeta(),outputRow);
+            }
+            return true;
+        } else {
+            throw new KettleException("Expected field " + meta.getInputField() + " to contain XML as type java.lang.String, but found "
+                    + xmlFieldValue.getClass());
+        }
+    }
+
+    private CanonicalizationResult process(String xmlString, CanonicalStepData data) {
+        try{
+            final Document xmlDoc = createDocument(xmlString,data);
+            return new CanonicalizationResult(canonicalize(xmlDoc,data));
+        } catch (KettleException kex) {
+            CanonicalizationResult result = new CanonicalizationResult("");
+            result.setErrorMessage(kex.getMessage());
+            result.setHasErrors(true);
+            result.incrementErrorCount();
+            return result;
+        }
+    }
+
+    private Canonicalizer getCanonicalizer() throws KettleException {
+        try {
+            return Canonicalizer.getInstance(Canonicalizer.ALGO_ID_C14N_EXCL_WITH_COMMENTS);
+        } catch (Exception e) {
+            throw new KettleException(e.getMessage(),e);
+        }
+    }
+
+    private DocumentBuilder getDocumentBuilder() throws KettleException {
+        final DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
```

Comment:
I *think* (but maybe check) that `DocumentBuilderFactory` can be static - as you can safely share an instance - and then lateral  call `newDocumentBuilder` on it where needed